### PR TITLE
refactor(pipeline): impl cluster info provider

### PR DESCRIPTION
**Pull request by chengjoey**
#### What this PR does / why we need it:
impl cluster info provider
make cache for clusterinfo

#### Which issue(s) this PR fixes:

- Fixes #your-issue_number
- [Erda Cloud Issue Link](https://erda-org.erda.cloud/erda/dop/projects/387/issues/gantt?filter__urlQuery=eyJhc3NpZ25lZSI6WyIxMDAxMjA1Il19&id=291252&iterationID=1146&pId=290591&type=TASK)

#### Specified Reviewers:

/assign @your-reviewer

#### ChangeLog

| Language | Changelog |
| --------- | ------------ |
| 🇺🇸 English |   Support pipeline clusterinfo provider           |
| 🇨🇳 中文    |  抽象pipeline的clusterinfo模块，统一处理            |

#### Need cherry-pick to release versions?

Add comment like `/cherry-pick release/1.0` when this PR is merged.

> For details on the cherry pick process, see the [cherry pick requests](https://github.com/erda-project/erda/blob/master/CONTRIBUTING.md#how-to-cherry-pick-a-merged-pr) section under [CONTRIBUTING.md](https://github.com/erda-project/erda/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: modules/pipeline/providers/clusterinfo/provider.go [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		default:
+			err := p.registerClusterHook()
+			if err == nil {
+				return
+			}
+			p.Log.Errorf("Failed to register cluster hook(auto retry), err: %v", err)
```
Comment (sfwn) [inline modules/pipeline/providers/clusterinfo/provider.go:59]: failed to

File: modules/pipeline/providers/engine/provider.go [hunk 2]
```diff
@@ -63,41 +62,20 @@ func (p *provider) Init(ctx servicehub.Context) error {
...
-
 	// action executor manager
 	_, cfgChan, err := p.dbClient.ListPipelineConfigsOfActionExecutor()
 	if err != nil {
 		return err
 	}
 	mgr := actionexecutor.GetManager()
 	p.actionExecutorMgr = mgr
-	if err := mgr.Initialize(ctx, cfgChan); err != nil {
+	if err := mgr.Initialize(ctx, cfgChan, p.ClusterInfo); err != nil {
```
Comment (sfwn) [inline modules/pipeline/providers/engine/provider.go:72]: I think action executor manager should be provider later :)

File: modules/pipeline/providers/reconciler/legacy/reconciler/define.go [hunk 3]
```diff
@@ -72,14 +74,14 @@ type PipelineSvcFunc struct {
 func New(js jsonstore.JsonStore, etcd *etcd.Store, bdl *bundle.Bundle, dbClient *dbclient.Client,
 	actionAgentSvc *actionagentsvc.ActionAgentSvc,
 	extMarketSvc *extmarketsvc.ExtMarketSvc,
-	pipelineSvcFunc *PipelineSvcFunc,
-	DBGC dbgc.Interface,
-) (*Reconciler, error) {
+	pipelineSvcFunc *PipelineSvcFunc, DBGC dbgc.Interface,
+	ci clusterinfo.Interface) (*Reconciler, error) {
```
Comment (sfwn) [inline modules/pipeline/providers/reconciler/legacy/reconciler/define.go:78]: no `ci`, just `clusterInfo`

**Review by sfwn — CHANGES_REQUESTED** [review]

File: modules/pipeline/providers/clusterinfo/cache.go [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+	c.cache[clusterInfo.Name] = clusterInfo
+}
+
+func (c *ClusterInfoCache) DeleteClusterInfo(name string) {
+	c.Lock()
+	defer c.Unlock()
+	delete(c.cache, name)
+}
+
+func (c *ClusterInfoCache) GetClusterInfoDuplication() []apistructs.ClusterInfo {
```
[the remark this answers is not in the document]
Comment (chengjoey) [reply modules/pipeline/providers/clusterinfo/cache.go:66]: i mean duplication of all clusters in cache

[the remark this answers is not in the document]
Comment (chengjoey) [reply modules/pipeline/providers/clusterinfo/cache.go:66]: rename to `GetAllClusters`

File: modules/pipeline/providers/clusterinfo/cache.go [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+package clusterinfo
+
+import (
+	"sync"
+
+	"github.com/erda-project/erda/apistructs"
+)
+
+type Cache interface {
```
[the remark this answers is not in the document]
Comment (chengjoey) [reply modules/pipeline/providers/clusterinfo/cache.go:23]: adjusted

File: modules/pipeline/providers/clusterinfo/interface.go [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+	"github.com/sirupsen/logrus"
+
+	"github.com/erda-project/erda/apistructs"
+	"github.com/erda-project/erda/bundle"
+	"github.com/erda-project/erda/pkg/discover"
+	"github.com/erda-project/erda/pkg/strutil"
+)
+
+var (
+	ClusterHookApiPath = "/api/pipeline-clusters/actions/hook"
```
[the remark this answers is not in the document]
Comment (chengjoey) [reply modules/pipeline/providers/clusterinfo/interface.go:29]: adjusted

File: modules/pipeline/providers/clusterinfo/interface.go [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+)
+
+type Interface interface {
+	GetClusterByName(string) (apistructs.ClusterInfo, error)
```
[the remark this answers is not in the document]
Comment (chengjoey) [reply modules/pipeline/providers/clusterinfo/interface.go:33]: added

File: modules/pipeline/providers/clusterinfo/provider.go (hunk 1 above)
Comment (chengjoey) [reply modules/pipeline/providers/clusterinfo/provider.go:59]: adjusted

File: modules/pipeline/providers/reconciler/legacy/reconciler/define.go (hunk 3 above)
Comment (chengjoey) [reply modules/pipeline/providers/reconciler/legacy/reconciler/define.go:78]: adjusted all places used as `ci`

File: modules/pipeline/providers/engine/provider.go (hunk 2 above)
Comment (chengjoey) [reply modules/pipeline/providers/engine/provider.go:72]: maybe should action executor should be provider
could create task for it

File: modules/pipeline/pipengine/actionexecutor/logic/job.go [hunk 8]
```diff
@@ -246,27 +251,20 @@ func GetBigDataConf(task *spec.PipelineTask) (apistructs.BigdataConf, error) {
...
 	}
+	return cluster, nil
+}
 
-	statusCode := resp.StatusCode()
-	respBody := body.String()
-
-	if err := json.Unmarshal([]byte(respBody), &clusterInfoRes); err != nil {
-		return nil, errors.Errorf("get cluster info failed, statueCode: %d, err: %v", statusCode, err)
+func GetClusterInfoByName(clusterName string) (map[string]string, error) {
```
Comment (sfwn) [inline modules/pipeline/pipengine/actionexecutor/logic/job.go:262]: bad name.
`GetClusterByName` return `apistructs.ClusterInfo`, `GetClusterInfoByName` return `map[string]string`.

File: modules/pipeline/pipengine/actionexecutor/plugins/k8sspark/k8sspark.go [hunk 9]
```diff
@@ -52,7 +50,7 @@ func init() {
 		if err != nil {
 			return nil, err
 		}
-		cluster, err := clusterinfo.GetClusterByName(clusterName)
+		cluster, err := logic.GetClusterByName(clusterName)
```
Comment (sfwn) [inline modules/pipeline/pipengine/actionexecutor/plugins/k8sspark/k8sspark.go:53]: maybe no need wrap by `logic`.

File: modules/pipeline/providers/clusterinfo/interface.go [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+	"github.com/sirupsen/logrus"
+
+	"github.com/erda-project/erda/apistructs"
+	"github.com/erda-project/erda/bundle"
+	"github.com/erda-project/erda/pkg/discover"
+	"github.com/erda-project/erda/pkg/strutil"
+)
+
+var (
+	clusterHookApiPath = "/api/pipeline-clusters/actions/hook"
```
Comment (sfwn) [inline modules/pipeline/providers/clusterinfo/interface.go:29]: I think this provider can provide http api directly.
remove explicit declaration in endpoint.

File: modules/pipeline/providers/clusterinfo/interface.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (sfwn) [reply modules/pipeline/providers/clusterinfo/interface.go:43]: you should add comment above each method directly.

such as:

```go
type Interface interface {
  // GetClusterInfoByName xxx
  GetClusterInfoByName ...

  // ListAllClusters xxx
  ListAllClusters ...
}
```

**Review by sfwn — CHANGES_REQUESTED** [review]

File: modules/pipeline/pipengine/actionexecutor/logic/job.go (hunk 8 above)
Comment (chengjoey) [reply modules/pipeline/pipengine/actionexecutor/logic/job.go:262]: GetClusterByName --> GetClusterInfoByName
GetClusterInfoByName --> GetClusterInfoCMByName

File: modules/pipeline/pipengine/actionexecutor/plugins/k8sspark/k8sspark.go (hunk 9 above)
Comment (chengjoey) [reply modules/pipeline/pipengine/actionexecutor/plugins/k8sspark/k8sspark.go:53]: adjusted, 
image

File: modules/pipeline/providers/clusterinfo/interface.go (hunk 10 above)
Comment (chengjoey) [reply modules/pipeline/providers/clusterinfo/interface.go:29]: added todo

File: modules/pipeline/providers/clusterinfo/interface.go [hunk 11]
```diff
@@ -0,0 +1,185 @@
...
+var (
+	ClusterHookApiPath = "/api/pipeline-clusters/actions/hook"
+)
+
+// Interface defines the interface for the clusterinfo provider
+type Interface interface {
+	// GetClusterByName return ClusterInfo query by name, contain cluster connection info and config map data
+	GetClusterInfoByName(string) (apistructs.ClusterInfo, error)
+	// GetClusterInfoCMByName return config map data of cluster query by name, data is displayed as map
+	GetClusterInfoCMByName(string) (map[string]string, error)
```
Comment (sfwn) [inline modules/pipeline/providers/clusterinfo/interface.go:37]: merge `GetClusterInfoByName` and `GetClusterInfoCMByName` into one.

**Review by sfwn — CHANGES_REQUESTED** [review]

File: modules/pipeline/providers/clusterinfo/interface.go (hunk 11 above)
Comment (chengjoey) [reply modules/pipeline/providers/clusterinfo/interface.go]: adjusted

File: modules/pipeline/providers/clusterinfo/interface.go [hunk 12]
```diff
@@ -0,0 +1,179 @@
...
+type Interface interface {
+	// GetClusterByName return ClusterInfo query by name, contain cluster connection info and config map data
+	GetClusterInfoByName(string) (apistructs.ClusterInfo, error)
+	// ListAllClusterInfos return all clusters, contain cluster connection info and config map data
+	ListAllClusterInfos() ([]apistructs.ClusterInfo, error)
+	// DispatchClusterEvent dispatch cluster event received from eventbox
+	// other modules can get event by channel if registered
+	DispatchClusterEvent(apistructs.ClusterEvent)
+	// GetClusterInfoDataByName return the cluster config map data, and transfer data to ClusterInfoData struct
+	GetClusterInfoDataByName(string) (apistructs.ClusterInfoData, error)
```
Comment (sfwn) [inline modules/pipeline/providers/clusterinfo/interface.go:42]: merge `GetClusterInfoByName` and `GetClusterInfoDataByName` into one.

**Review by sfwn — CHANGES_REQUESTED** [review]

**Review by sfwn — APPROVED** [review]

**Pull request merged**